Modern/ops/src/main/java/org/bonej/ops/ellipsoid/FindEllipsoidFromBoundaryPoints.java

```diff
@@ -0,0 +1,343 @@
+ * <p>
+ * The ellipsoid decomposition is drawn from the paper of Bischoff and Kobbelt, 2002 (https://www.graphics.rwth-aachen.de/media/papers/ellipsoids1.pdf)
+ * The variable naming widely follows Bischoff and Kobbelt's nomenclature, it is recommended to read this code in conjunction with the paper.
+ * The Ellipsoid-Plane intersection is done slightly differently, using the {@link EllipsoidPlaneIntersection}. This op starts growing an ellipsoid
+ * from the starting point p, and keeps growing in various directions until it hits three more points (q,r,s). If less than four distinct points are found,
+ * the op returns an empty ellipsoid. More details on the growing process are in the original publication.
+ * </p>
+ *
+ * @author Alessandro Felder
+ */
+@Plugin(type = Op.class)
+public class FindEllipsoidFromBoundaryPoints extends AbstractBinaryFunctionOp<List<ValuePair<Vector3d,Vector3d>>,Vector3d,Optional<Ellipsoid>> {
+
+    private static QuadricToEllipsoid quadricToEllipsoid = new QuadricToEllipsoid();
+    private static EllipsoidPlaneIntersection intersectionOp = new EllipsoidPlaneIntersection();
+
+    @Parameter(persist=false, required = false)
+    private double calibratedLargestImageDimension = 100.0;
+
+    @Override
+    public Optional<Ellipsoid>calculate(List<ValuePair<Vector3d,Vector3d>> fourVerticesWithNormals, Vector3d centre) {
+        List<Vector3d> fourVertices = new ArrayList<>();
+        fourVerticesWithNormals.forEach(v -> fourVertices.add(v.getA()));
+
+        fourVerticesWithNormals.sort(Comparator.comparingDouble(p->euclideanDistance(p.getA(),centre)));
+        VertexWithNormal p = new VertexWithNormal(fourVerticesWithNormals.get(0));
+        fourVertices.remove(p.getVertex());
+
+        ValuePair<VertexWithNormal, Double> qAndRadius = calculateQ(fourVertices, p);
+        if (qAndRadius == null || qAndRadius.getB()>calibratedLargestImageDimension) return Optional.empty();
+        fourVertices.remove(qAndRadius.getA().getVertex());
+
+        Vector3d np = new Vector3d(p.getNormal());
+        np.mul(qAndRadius.getB());
+        p.setNormal(np);
+
+        Vector3d c = new Vector3d(p.getNormal());
+        c.add(p.getVertex());
+        Matrix4d q1 = getQ1(c, qAndRadius.getB());
+        Matrix4d q2 = getQ2(p, qAndRadius.getA());
+
+        ValuePair<Vector3d, Double> rAndAlpha = calculateSurfacePointAndGreekCoefficient(q1, q2, fourVertices);
+        if (rAndAlpha == null) return Optional.empty();
+        fourVertices.remove(rAndAlpha.getA());
+
+        Matrix4d q1PlusAlphaQ2 = new Matrix4d(q2);
+        Matrix4d fullScalingMatrix = new Matrix4d();
+        fullScalingMatrix.scaling(rAndAlpha.getB());
+        fullScalingMatrix.m33(rAndAlpha.getB());
+        q1PlusAlphaQ2.mul(fullScalingMatrix, q1PlusAlphaQ2);
+        q1PlusAlphaQ2.add(q1, q1PlusAlphaQ2);
+
+        Optional<Ellipsoid> ellipsoid = quadricToEllipsoid.calculate(q1PlusAlphaQ2);
```

Comment:
Replace lines _57 -- 68_ with a helper method `Optional<Ellipsoid> ellipsoid = createEllipsoid(q1, q2)` (you can probably think of a better name)